### PR TITLE
Remove bower and port dependencies over to npm

### PR DESCRIPTION
**Pull request by petrgazarov**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Chore

#### Checklist:

- [x] Tested changes locally.
- [x] Closes currently open issue: https://github.com/FreeCodeCamp/FreeCodeCamp/issues/11461

#### Description

- Removes bower
- Ports dependencies that were required through Bower to `npm`.
- Ads CDNs for CSS files

Some thoughts on future improvements (would love to hear what you guys think about these!):
- Replace `npm` with `yarn`
  - Super easy transition, for most cases
- Remove CDNs for libraries like `font-awesome`, `codemirror` and `bootstrap`. These can be loaded and served by webpack.
  - Right now CDNs for `font-awesome` alone are in 12+ places. And require two different versions, I can imagine that's not ideal to manage.

Appreciate review and any feedback/comments. Thanks!!

### CODE REVIEW

**Comment by atjonathan** [comment]
@petrgazarov, please open your suggestions in a separate issue 😄

**Comment by raisedadead** [comment]
@petrgazarov thanks a lot for this. It's a good move to standardise the dependencies to as smaller sets as possible.

Just a quick query:
I saw that the PR changes mousetrap to a lower version, was this intended? Any particular reason?

As for the suggestions lets take it up in a separate thread.

**Comment by petrgazarov** [comment]
@raisedadead Good point. `mousetrap` was a dependency in both bower and npm, but from what I could tell, only the one in bower (version `1.5.3`) was actually used.

File: package.json
Comment (BerkeleyTrue) [inline package.json]: Rx is on here twice. Are you manually editing this file? You should use npm to do this.

File: server/views/certificate/data-vis.jade
Comment (BerkeleyTrue) [inline server/views/certificate/data-vis.jade]: We don't rely on cdn's for core libs as they are often blocked in different countries.

File: server/views/redirect-https.html
Comment (BerkeleyTrue) [inline server/views/redirect-https.html]: This file does not need to be changed

File: client/commonFramework/update-preview.js
Comment (BerkeleyTrue) [inline client/commonFramework/update-preview.js]: '/node_modules' is not a public path

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

File: server/views/certificate/data-vis.jade
Comment (petrgazarov) [inline server/views/certificate/data-vis.jade]: Gtk. What do you think about configuring webpack to serve css files from `node_modules`, rather than use cdn.

Comment (BerkeleyTrue) [inline server/views/certificate/data-vis.jade]: > What do you think about configuring webpack to serve css files from node_modules

What do you mean by this?

Comment (petrgazarov) [inline server/views/certificate/data-vis.jade]: AFAIK, webpack can also be configured to serve CSS from node_modules by defining a css loader in config. See [this discussion for a more context](https://github.com/webpack/webpack/issues/1789). For example, bootstrap css files can make it to the app via webpack and a `require()` in a JS file. Thus no need in CDNs.

Apologies if I'm speaking out of context, since I'm new to the codebase, would like to hear your opinion.

File: package.json
Comment (petrgazarov) [inline package.json]: Yup, will fix this one. Not sure how it got in there. Good catch!

File: server/views/certificate/data-vis.jade
Comment (BerkeleyTrue) [inline server/views/certificate/data-vis.jade]: This would need to be a separate process from the main app.

File: server/views/certificate/front-end.jade
Comment (BerkeleyTrue) [inline server/views/certificate/front-end.jade]: I'd prefer cdnjs for this as we already use cloudflare.

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

File: gulpfile.js
Comment (BerkeleyTrue) [inline gulpfile.js]: The problem with this is that npm install is not deterministic.

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

File: gulpfile.js
Comment (petrgazarov) [inline gulpfile.js]: Hm. Interesting. So this can potentially break if higher version was installed on the top level, via dependency of other module, which is alphabetically before `codemirror`. [Shrinkwrap](https://docs.npmjs.com/cli/shrinkwrap) can be used to solve this and lock down versions. I've used it and the only downside is that you have to manually update it whenever you change your package.json.

**Comment by petrgazarov** [comment]
Created JQuery upgrade [separate issue](https://github.com/FreeCodeCamp/FreeCodeCamp/issues/12015).
Any chance for a re-review here? Would love to merge this if you are still interested. Thanks

**Comment by raisedadead** [comment]
thanks for the updates @petrgazarov , just waiting on @BerkeleyTrue for a review. sorry about the delay, he has got a lot on his hands, rest assured we will get back as soon as possible.

File: gulpfile.js
Comment (BerkeleyTrue) [inline gulpfile.js]: This is still a problem. I don't think shrinkwrap is a viable option. You could try using the path module to programmatically determine the routes to projects.

Comment (petrgazarov) [inline gulpfile.js]: @BerkeleyTrue Hm, maybe I'm missing something but I checked and neither `path` nor `require.resolve()` have functionality to solve for non-determinism that you've mentioned. Some observations:

- `require.resolve()` can be effective when calling from code, but when called from top scope, it just returns the global node_modules path for the module if it finds one.
- `path.resolve()` can check if a path exists in multiple sub-directories, but we don't know where npm decides to install the exact version we need, so it won't help.

I don't think there is a way to go around non-determinism with npm, if not use shrinkwrap. For that matter yarn can help and [I've raised issue to consider migration to yarn](https://github.com/FreeCodeCamp/FreeCodeCamp/issues/11501).

An alternative solution would be using cdns in place of all the `js` files that are being complied into `vendor-challenges.js` and `vendor-main.js`, but that may not be desirable for cdn accessibility reasons you mentioned and also increasing the number of network requests significantly.

I'll let it up to you to make the strategic long-term call on this, but happy to help with a PR if you decide on the approach.

Or, if there is an potential solution that I have not listed, would be happy to weight in on it as well.

Comment (BerkeleyTrue) [inline gulpfile.js]: > require.resolve() can be effective when calling from code, but when called from top scope, it just returns the global node_modules path for the module if it finds one.

I'm not sure what you mean by that statement.

> require.resolve('react')
'/Users/berkeley/node/rwr/actual/node_modules/react/react.js'
> require.resolve('react-router')
'/Users/berkeley/node/rwr/actual/node_modules/react-router/lib/index.js'
> require.resolve('warning')
'/Users/berkeley/node/rwr/actual/node_modules/warning/warning.js'

Here I use the `require.resolve` method to deliver the directory of the main file for a module according to the rules set by node. 

This can be used to find the directory relative to `node_modules`. With a little string concatenation magic we have the correct path to the correct file.

Comment (petrgazarov) [inline gulpfile.js]: Yup, that seems right. However:

`require.resolve` relies on [lookup order used by `require`](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders). If you are calling `require.resolve` in gulpfile, the first directory it looks in will be the global `node_modules` for the project.

For example, 
`require.resolve('jquery)` in gulpfile will return path for the jquery version that lives in `FreeCodeCamp/node_modules/jquery/`

So these two lines point to the same file when we write it in gulpfile:
`'node_modules/jquery/dist/jquery.js'`
and
`require.resolve('jquery')`

To your earlier point, npm3 is non-deterministic, and `FreeCodeCamp/node_modules/jquery/` may potentially contain the wrong version of `jquery`. That's basically why I'm advocating for  using shrinkwrap with npm. Shrinkwrap locks the dependencies of dependencies, so `FreeCodeCamp/node_modules/jquery/` will always contain the same jquery version regardless on the environment.

Comment (BerkeleyTrue) [inline gulpfile.js]: > To your earlier point, npm3 is non-deterministic, and FreeCodeCamp/node_modules/jquery/ may potentially contain the wrong version of jquery

Yes, but `require.resolve` should always point to the one listed in package.json.
`require.resolve` keeps us from pulling the wrong version.

> That's basically why I'm advocating for using shrinkwrap with npm

We won't be moving to Shrinkwrap just for this issue.

**Review by BerkeleyTrue — APPROVED** [review]

**Comment by BerkeleyTrue** [comment]
Unfortunately it looks like there are security issues with jquery 2.X

I'm looking into the upgrade path to 3.0.

**Comment by BerkeleyTrue** [comment]
Looks like there are minimal api changes between 2.X and 3.0. Do you mind updating the $ dependencies to 3.0?

**Comment by petrgazarov** [comment]
Yeah! Hopefully I can crave out some time this weekend.

**Comment by BerkeleyTrue** [comment]
@petrgazarov Are you still able to make that change?

**Comment by petrgazarov** [comment]
@BerkeleyTrue Got caught up with a recent hackathon, sorry. I will have some time to work on it this week, and if the upgrade is not very involved, even finish by Thursday-Friday. If that's too late, feel free to re-assign. Apologies for the delay!

**Comment by petrgazarov** [comment]
PR for upgrading JQuery: https://github.com/freeCodeCamp/freeCodeCamp/pull/12605

That one should probably be merged first, to make Snyk happy.

**Comment by BerkeleyTrue** [comment]
@petrgazarov Awesome work so far. Do you mind doing a rebase?

**Comment by petrgazarov** [comment]
Hey @BerkeleyTrue, thanks! I rebased.

The app loads the main pages, I wasn't able, however, to get to challenges pages this time to test them. The app requires sign in, and each method that I've tried fails locally. Perhaps I don't have the required env variables?

Is there a recommended way to set up the db with a test user that gives access to these pages for testing purposes?

**Comment by petrgazarov** [comment]
Rebased branch :1st_place_medal:

**Comment by raisedadead** [comment]
> The app loads the main pages, I wasn't able, however, to get to challenges pages this time to test them. The app requires sign in, and each method that I've tried fails locally. Perhaps I don't have the required env variables?

For GitHub:
Yes, you need to register an application in https://github.com/settings/developers
with the callback URL `http://localhost:3000/auth/github/callback`, then add the keys in `.env`

```
GITHUB_ID=xxxxxxxxxxxxxxxxxxxx
GITHUB_SECRET=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

For Email:
You configure Amazon SES and add the keys in `.env` or you could use the a temporary gmail config in the [`server/datasources.json`](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/server/datasources.json) 

Details are available here https://loopback.io/doc/en/lb3/Email-connector.html#using-gmail

On that note, you need not do the email, setup at all. If you try and signup on the signup page, the server will crash, because it could not send the email. But typing `rs` will restart the nodemon and you can continue to login as it is.

> Is there a recommended way to set up the db with a test user that gives access to these pages for testing purposes?

No, not currently but yeah, it would be great to have such a feature.

Please hit us up on the chat room if you need assistance, and thanks a ton for the contribution...! Have been playing with these changes, and so far its working great!

**Comment by BerkeleyTrue** [comment]
@petrgazarov Thanks!

Happy Coding

**Pull request merged**
